extension/oauth2authextension/extension.go

```diff
@@ -0,0 +1,90 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oauth2authextension
+
+import (
+	"context"
+	"net/http"
+
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc/credentials"
+	grpcOAuth "google.golang.org/grpc/credentials/oauth"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configauth"
+)
+
+// OAuth2Authenticator provides implementation for providing client authentication using OAuth2 client credentials
+// workflow for both gRPC and HTTP clients.
+type OAuth2Authenticator struct {
+	clientCredentials *clientcredentials.Config
```

Comment:
This is only implementing the client credentials flow. OAuth2 is a large spec and client credentials is not even the most implemented flow. I don't think we should dedicate the "oauth2" name to the client credentials flow.

Reply:
Thanks @rakyll  - I agree.  Does `OAuth2ClientCredentialsAuthenticator` makes sense? and also the factory type string is `oauth2clientcredentials`

```
extensions:
  oauth2clientcredentials:
    client_id: someclientid
    client_secret: someclientsecret
    token_url: https://someserver.com/oauth2/default/v1/token
    scopes: ["api.metrics"]

receivers:
  hostmetrics:
    scrapers:
      memory:
  otlp:
    protocols:
      grpc:

exporters:
  otlphttp/withauth:
    endpoint: http://localhost:9000
    auth:
      authenticator: oauth2clientcredentials
```

Reply:
Also @jpkrohling @rakyll the package name is getting really big and ugly - `oauth2clientcredentialsauthextension` - any suggestion? does `auth/oauth2/clientcredentialsextension` makes sense? (this assumes we repackage `oidcauthextension` and `bearertokenauthextension` under `auth` as well